lib/Conversion/XTenNNToTorch.cpp

```diff
@@ -0,0 +1,170 @@
+using namespace mlir::torch;
+
+namespace {
+
+Type getCompatibleTorchDType(MLIRContext *ctx, Type dtype) {
+  if (isa<FloatType>(dtype))
+    return dtype;
+
+  auto integerType = dtype.cast<IntegerType>();
+  if (!integerType.isSignless())
+    return dtype;
+
+  // Torch builtin types only support Signed and Unsigned integer types.
+  // Convert Signless to Signed.
+  return IntegerType::get(ctx, integerType.getWidth(),
+                          mlir::IntegerType::Signed);
+}
+
+Value toTorchTensorTypeCast(PatternRewriter &rewriter, Value input) {
+
+  auto tensorTy = dyn_cast<ShapedType>(input.getType());
+  auto sizes = tensorTy.getShape();
+
+  auto dtype =
+      getCompatibleTorchDType(rewriter.getContext(), tensorTy.getElementType());
+
+  return rewriter
+      .create<TorchConversion::FromBuiltinTensorOp>(
+          input.getLoc(),
+          mlir::torch::Torch::ValueTensorType::get(input.getContext(), sizes,
+                                                   dtype),
+          input)
+      .getResult();
+}
+
+Value toBuiltinTensorTypeCast(OpBuilder &builder, Value val, Type type) {
+  if (val.getType().isa<MemRefType>())
+    return val;
+
+  auto tensorTy = val.getType().dyn_cast<torch::Torch::BaseTensorType>();
+  if (!tensorTy)
+    return val;
+  return builder.create<torch::TorchConversion::ToBuiltinTensorOp>(val.getLoc(),
+                                                                   type, val);
+}
+
+template <typename... ops>
+class GenericPatternXTenNNToTorch : public RewritePattern {
+public:
+  GenericPatternXTenNNToTorch(MLIRContext *ctx)
+      : RewritePattern(MatchAnyOpTypeTag(), /*benefit=*/1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op,
+                                PatternRewriter &rewriter) const override {
+
+    auto *ctx = op->getContext();
+    if (isa<ops...>(op)) {
+      return rewriter.notifyMatchFailure(op->getLoc(),
+                                         "no conversion for this operation.");
+    }
+
+    if (op->getName().getDialectNamespace() != "xten_nn") {
+      return rewriter.notifyMatchFailure(
+          op->getLoc(), "operation doesn't belong to XTenNN dialect.");
+    }
+
+    SmallVector<Value> vtensorOperands;
+    llvm::transform(
+        op->getOperands(), std::back_inserter(vtensorOperands),
+        [&](Value val) { return toTorchTensorTypeCast(rewriter, val); });
+
+    // Convert MLIR types to Torch builtin types.
+    SmallVector<Type> vtensorResultTypes;
+    llvm::transform(
+        op->getResultTypes(), std::back_inserter(vtensorResultTypes),
+        [&](Type ty) {
+          auto tensorTy = cast<TensorType>(ty);
+          return Torch::ValueTensorType::get(
+              ctx, tensorTy.getShape(),
+              getCompatibleTorchDType(ctx, tensorTy.getElementType()));
+        });
+
+    // Start composing new op
+    OperationState state(
+        op->getLoc(), "torch.aten." + std::string(op->getName().stripDialect()),
```

Comment:
This imposes the implicit restriction of xten operations having the same name than the torch counterpart. If this requirement is not meet we will generate invalid torch and the same applies to for attributes.

My concern about this generic rewriting is that if we introduce a new op (as I did with GroupConv) then this pass will be invalid automatically because this generic rewriting.

What do you think of creating a base pattern where common parts are performed (conversion of tensor operands and results) and we call one function for each pattern we want to translate. This way new ops won't invalidate this pass.

Reply:
Done!